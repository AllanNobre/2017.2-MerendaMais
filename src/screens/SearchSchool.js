import axios from 'axios';
import React from 'react';
import { StyleSheet, Text, View, TouchableOpacity, Image, TextInput, FlatList, ActivityIndicator, Alert } from 'react-native';
import { SideMenu } from 'react-native-elements';
import { Actions } from 'react-native-router-flux';
import Menu from '../components/Menu';

const sideMenuIcon = require('../images/ic_menu_black_48dp_1x.png');
const CityIcon = require('../images/ic_location_city_48pt.png');
const SearchIcon = require('../images/ic_search_48pt.png');
const BackIcon = require('../images/ic_keyboard_arrow_left_48pt.png');
const GoIcon = require('../images/ic_keyboard_arrow_right_48pt.png');


const styles = StyleSheet.create({
  headerBox: {
    flex: 1,
    padding: 10,
    backgroundColor: '#FF9500',
    borderBottomColor: 'black',
    borderBottomWidth: 1,
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center',
  },
  bodyBox: {
    flex: 10,
    alignItems: 'center',
  },
  textLogo: {
    flex: 1,
    fontSize: 30,
    color: 'white',
    fontWeight: 'bold',
    textAlign: 'center',
  },
  Input: {
    marginTop: 20,
    paddingLeft: 2,
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center',
    borderColor: 'gray',
    backgroundColor: '#FAFAFA',
    borderWidth: 1,
    borderRadius: 7,
  },
  icon: {
    width: 30,
    height: 30,
    margin: 5,
  },
  buttonSearchAnabled: {
    paddingHorizontal: 117,
    paddingVertical: 15,
    marginTop: 50,
    marginBottom: 0,
    backgroundColor: '#FF9500',
    borderRadius: 8,
    borderWidth: 1,
  },
  buttonSearchDisabled: {
    paddingHorizontal: 117,
    paddingVertical: 15,
    marginTop: 50,
    marginBottom: 0,
    backgroundColor: '#DEDEDE',
    borderRadius: 8,
    borderWidth: 1,
  },
  listSchools: {
    flex: 5,
    justifyContent: 'center',
    width: 320,
    borderColor: 'black',
    borderWidth: 1,
    borderRadius: 7,
  },
  item: {
    borderBottomColor: 'black',
    borderBottomWidth: 0.5,
    borderRadius: 7,
  },
  buttonSelectSchool: { padding: 10,
    alignItems: 'center',
    flexDirection: 'row',
    justifyContent: 'space-between',
  },
  buttonArea: {
    flex: 1,
    justifyContent: 'flex-end',
    paddingBottom: 20,
  },
});

console.disableYellowBox = true;

class SearchSchool extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      isOpen: false,
      isLoading: false,
      city: '',
      name: '',
      schoolList: [],
    };
    this.validateName = this.validateName.bind(this);
    this.validateCity = this.validateCity.bind(this);
  }
  validateName(name) {
    const validName = name.replace(/[^A-Za-záàâãéèêíïóôõöúçñÁÀÂÃÉÈÍÏÓÔÕÖÚÇÑ0-9 ]/g, '');
    this.setState({ name: validName });
  }

  validateCity(city) {
    const validCity = city.replace(/[^A-Za-záàâãéèêíïóôõöúçñÁÀÂÃÉÈÍÏÓÔÕÖÚÇÑ ]/g, '');
    this.setState({ city: validCity });
  }

  register() {
    const nameRegex = /[A-Za-záàâãéèêíïóôõöúçñÁÀÂÃÉÈÍÏÓÔÕÖÚÇÑ0-9 ]/g;
    const cityRegex = /[A-Za-záàâãéèêíïóôõöúçñÁÀÂÃÉÈÍÏÓÔÕÖÚÇÑ ]/g;
    let error = false;
    let errorMessage = '';

    if (!cityRegex.test(this.state.city) && !nameRegex.test(this.state.name)) {
      error = true;
      errorMessage += 'Estado/Município e Escola com dados inválidos.';
    }
    if (this.state.city.trim() === '' && this.state.name.trim() === '') {
      error = true;
      errorMessage += 'Estado/Município e Escola não preenchidos. Preencha no mínimo um dos campos.\n';
    }
    if (error === false) {
      this.searchSchools();
    } else {
      Alert.alert('FALHA NA PESQUISA', errorMessage);
    }
  }

  searchSchools() {
    const params = {
      name: this.state.name,
      city: this.state.city,
    };
    this.setState({ isLoading: true });

    axios.post('https://merenda-mais.herokuapp.com/get_schools/', params)
      .then((response) => {
        console.log(response.data);
        this.setState({ schoolList: response.data, isLoading: false });

        console.log('SearchSchool State');
        console.log(this.state);
        // If response is an empty array, no schools could be found.
      })
      .catch((error) => {
        this.setState({ isLoading: false });
        console.log(error);
      });
  }

  updateMenuState(isOpen) {
    this.setState({ isOpen });
  }

  buttonActivation() {
    console.log(this.state.city);
    if (this.state.city > '' || this.state.name > '') {
      if (this.state.isLoading) {
        return (
          <ActivityIndicator
            key="Show_ActivityIndicator"
            style={{ paddingBottom: 10 }}
            size="large"
            color="#FF9500"
          />
        );
      }
      return (
        <TouchableOpacity
          key="Show_TouchableOpacity"
          style={styles.buttonSearchAnabled}
          activeOpacity={0.7}
          onPress={() => this.register()}
        >
          <Text style={{ color: 'white', fontSize: 20 }}>Pesquisar</Text>
        </TouchableOpacity>
      );
    }
    return (
      <TouchableOpacity
        style={styles.buttonSearchDisabled}
        disabled
      >
        <Text style={{ color: 'white', fontSize: 20 }}>Pesquisar</Text>
      </TouchableOpacity>
    );
  }

  render() {
    const menu = <Menu />;

    return (
      <SideMenu
        menu={menu}
        menuPosition="right"
        isOpen={this.state.isOpen}
        disableGestures
        onChange={isOpen => this.updateMenuState(isOpen)}
      >
        <View style={{ flex: 1, backgroundColor: 'white' }}>
          <View style={styles.headerBox}>
            <TouchableOpacity
              onPress={() => Actions.pop()}
            >
              <Image source={BackIcon} />
            </TouchableOpacity>
            <Text style={styles.textLogo}>Pesquisar Escola</Text>
            <TouchableOpacity
              onPress={() => this.updateMenuState(true)}
            >
              <Image source={sideMenuIcon} />
            </TouchableOpacity>
          </View>
          <View style={styles.bodyBox}>
            <View style={{ flex: 2 }}>
              <View style={styles.Input}>
                <Image source={CityIcon} style={styles.icon} />
                <TextInput
                  width={280}
                  returnKeyType="go"
                  maxLength={50}
                  keyboardType={'default'}
                  onChangeText={text => this.validateCity(text)}
                  value={this.state.city} // CAE OR CITY ?
                  underlineColorAndroid="transparent"
                  placeholder="Estado/Município"
                />
              </View>

              <View style={styles.Input}>
                <Image source={SearchIcon} style={styles.icon} />
                <TextInput
                  width={280}
                  returnKeyType="go"
                  maxLength={50}
                  keyboardType={'default'}
                  onChangeText={text => this.validateName(text)}
                  value={this.state.name}
                  underlineColorAndroid="transparent"
                  placeholder="Escola a pesquisar"
                />
              </View>
            </View>

            <View style={styles.listSchools} >
              <FlatList
                data={this.state.schoolList}
                keyExtractor={item => item.nome}
                renderItem={({ item }) => (
                  <View style={styles.item}>
                    <TouchableOpacity
                      style={styles.buttonSelectSchool}
                    >
                      <Text style={{ fontSize: 16 }}>{item.nome}</Text>
                      <Image source={GoIcon} style={{ width: 25, height: 25 }} />
                    </TouchableOpacity>
                  </View>
                )}
              />
            </View>

            <View style={styles.buttonArea}>
              {this.buttonActivation()}
            </View>
          </View>
        </View>
      </SideMenu>
    );
  }
}

export default SearchSchool;
