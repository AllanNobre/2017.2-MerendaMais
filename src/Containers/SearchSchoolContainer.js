import { connect } from 'react-redux';
import SearchSchool from '../screens/SearchSchool';
import setSchoolInfo from '../actions/schoolActions';

const mapStateToProps = state => ({
  counselor: state.counselor,
});

const mapDispatchToProps = dispatch => ({
  setSchoolInfo: selectedSchoolCode => dispatch(setSchoolInfo({ schoolCode: selectedSchoolCode })),
});

const SearchSchoolContainer = connect(mapStateToProps, mapDispatchToProps)(SearchSchool);

export default SearchSchoolContainer;
