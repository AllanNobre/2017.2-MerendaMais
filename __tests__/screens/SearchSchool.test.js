import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import configureStore from 'redux-mock-store';
import SearchSchoolContainer from '../../src/Containers/SearchSchoolContainer';
import SearchSchool from '../../src/screens/SearchSchool';

Enzyme.configure({ adapter: new Adapter() });

const mockStore = configureStore();

const initialState = {
  application: {
    isLoading: false,
    message_erro: '',
  },
};

const store = mockStore(initialState);

describe('Testing SearchSchool', () => {
  it('renders as expected', () => {
    const wrapper = shallow(
      <SearchSchoolContainer />,
      { context: { store } },
    ).dive();
    expect(wrapper).toMatchSnapshot();
  });
});

describe('Testing SearchSchool Input', () => {
  const wrapper = shallow(<SearchSchool />);

  it('should change state when the text of city input component changes', () => {
    const cityInputComponent = wrapper.find('TextInput').at(0);
    cityInputComponent.simulate('ChangeText', 'Gama');
    expect(wrapper.state('city')).toEqual('Gama');
  });

  it('should change state when the text of name input component changes', () => {
    const nameInputComponent = wrapper.find('TextInput').at(1);
    nameInputComponent.simulate('ChangeText', 'FGA');
    expect(wrapper.state('name')).toEqual('FGA');
  });

  it('should change state when the uf changes', () => {
    const ufInputComponent = wrapper.find('Picker').at(0);
    ufInputComponent.simulate('valueChange', 'DF');
    expect(wrapper.state('uf')).toEqual('DF');
  });
});

describe('Testing SearchSchool button', () => {
  it('Test if SearchSchool Button is rendered', () => {
    const wrapper = shallow(<SearchSchool />);
    const button = wrapper.findWhere(c => c.key() === 'renderButton');
    expect(button.length).toEqual(1);
    button.simulate('press');
  });
});

describe('Testing SearchSchool register method', () => {
  it('Test if validation is called', () => {
    const wrapper = shallow(<SearchSchool />);
    const spy = jest.spyOn(SearchSchool.prototype, 'register');

    wrapper.setState({ name: '#' });
    wrapper.instance().register();

    expect(spy).toHaveBeenCalled();
    spy.mockClear();
  });

  it('Test if validation for empty fields is called', () => {
    const wrapper = shallow(<SearchSchool />);
    const spy = jest.spyOn(SearchSchool.prototype, 'register');

    wrapper.setState({ name: '' });
    wrapper.instance().register();

    expect(spy).toHaveBeenCalled();
    spy.mockClear();
  });

  it('Test if validation for correct fields is called', () => {
    const wrapper = shallow(<SearchSchool />);
    const spy = jest.spyOn(SearchSchool.prototype, 'register');

    wrapper.setState({ city: 'Porto Velho' });
    wrapper.instance().register();

    expect(spy).toHaveBeenCalled();
    spy.mockClear();
  });

  /* eslint-disable no-param-reassign */
  it('Test if wrong input is validated', () => {
    let error = false;
    let errorMessage = '';
    const wrapper = shallow(<SearchSchool />);

    const spy = jest.spyOn(SearchSchool.prototype, 'register').mockImplementation(
      () => {
        const nameRegex = /[A-Za-záàâãéèêíïóôõöúçñÁÀÂÃÉÈÍÏÓÔÕÖÚÇÑ0-9 ]/g;
        const cityRegex = /[A-Za-záàâãéèêíïóôõöúçñÁÀÂÃÉÈÍÏÓÔÕÖÚÇÑ ]/g;

        if (!cityRegex.test(wrapper.state('city') && !nameRegex.test(wrapper.state('name')) && !(wrapper.state('uf') > ''))) {
          error = true;
          errorMessage += 'Estado/Município e Escola com dados inválidos.';
        }
        if (wrapper.state('city').trim() === '' && wrapper.state('name').trim() === '' && wrapper.state('uf').trim() === '') {
          error = true;
          errorMessage += 'Estado/Município e Escola não preenchidos. Preencha no mínimo um dos campos.\n';
        }
      });


    wrapper.setState({ name: '#' });

    wrapper.instance().register();

    expect(spy).toHaveBeenCalled();
    expect(error).toBe(true);
    expect(errorMessage).toBe('Estado/Município e Escola com dados inválidos.');
    spy.mockClear();
  });
  /* eslint no-param-reassign: 0 */
});
