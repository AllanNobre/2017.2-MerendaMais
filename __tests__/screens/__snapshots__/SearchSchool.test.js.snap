// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing SearchSchool renders as expected 1`] = `
<View
  style={
    Object {
      "backgroundColor": "white",
      "flex": 1,
    }
  }
>
  <Header
    backButton={true}
    subTitle=""
    title="Pesquisar Escola"
  />
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 10,
      }
    }
  >
    <View
      style={
        Object {
          "flex": 3,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#FAFAFA",
            "borderColor": "gray",
            "borderRadius": 7,
            "borderWidth": 1,
            "flexDirection": "row",
            "justifyContent": "center",
            "marginTop": 20,
            "paddingLeft": 2,
          }
        }
      >
        <Icon
          allowFontScaling={false}
          color="black"
          name="location-city"
          size={32}
          style={
            Object {
              "margin": 8,
            }
          }
        />
        <TextInput
          keyboardType="default"
          maxLength={50}
          onChangeText={[Function]}
          placeholder="Estado/MunicÃ­pio"
          returnKeyType="go"
          underlineColorAndroid="transparent"
          value=""
          width={280}
        />
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#FAFAFA",
            "borderColor": "gray",
            "borderRadius": 7,
            "borderWidth": 1,
            "flexDirection": "row",
            "justifyContent": "center",
            "marginTop": 20,
            "paddingLeft": 2,
          }
        }
      >
        <Icon
          allowFontScaling={false}
          color="black"
          name="search"
          size={30}
          style={
            Object {
              "margin": 8,
            }
          }
        />
        <TextInput
          keyboardType="default"
          maxLength={50}
          onChangeText={[Function]}
          placeholder="Escola a pesquisar"
          returnKeyType="go"
          underlineColorAndroid="transparent"
          value=""
          width={280}
        />
      </View>
      <View
        style={
          Object {
            "borderColor": "gray",
            "borderRadius": 7,
            "borderWidth": 1,
            "marginTop": 15,
            "paddingLeft": 2,
          }
        }
      >
        <Picker
          mode="dialog"
          onValueChange={[Function]}
          selectedValue=""
        >
          <PickerItem
            color="#95a5a6"
            label="Escolha sua UF"
            value=""
          />
          <PickerItem
            label="DF"
            value="DF"
          />
        </Picker>
      </View>
    </View>
    <View
      style={
        Object {
          "borderColor": "black",
          "borderRadius": 7,
          "borderWidth": 1,
          "flex": 2.5,
          "justifyContent": "center",
          "width": 320,
        }
      }
    >
      <FlatList
        data={Array []}
        disableVirtualization={false}
        horizontal={false}
        initialNumToRender={10}
        keyExtractor={[Function]}
        maxToRenderPerBatch={10}
        numColumns={1}
        onEndReachedThreshold={2}
        renderItem={[Function]}
        scrollEventThrottle={50}
        updateCellsBatchingPeriod={50}
        windowSize={21}
      />
    </View>
    <View
      key="renderButton"
      style={
        Object {
          "flex": 1,
          "justifyContent": "flex-end",
          "paddingBottom": 20,
        }
      }
    >
      <TouchableOpacity
        activeOpacity={0.2}
        disabled={true}
        style={
          Object {
            "backgroundColor": "#DEDEDE",
            "borderRadius": 8,
            "borderWidth": 1,
            "marginBottom": 0,
            "marginTop": 50,
            "paddingHorizontal": 117,
            "paddingVertical": 15,
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          disabled={false}
          ellipsizeMode="tail"
          style={
            Object {
              "color": "white",
              "fontSize": 20,
            }
          }
        >
          Pesquisar
        </Text>
      </TouchableOpacity>
    </View>
  </View>
</View>
`;
